File: chap11/fig11-9.py
import pandas
import cmdstanpy
import numpy as np
from plotnine import ggplot, aes, theme_bw, geom_ribbon, geom_line, labs

d = pandas.read_csv('input/data-switch.csv')
T = len(d)

fit = cmdstanpy.from_csv('output/result-model11-6')
sw_ms = fit.stan_variable(var='sw')
qua = np.quantile(sw_ms, [0.025, 0.25, 0.50, 0.75, 0.975], axis=0)
d_est = pandas.DataFrame(np.column_stack([np.arange(1, T+1), qua.T]), \
    columns=['X', '2.5%', '25%', '50%', '75%', '97.5%'])

p = (ggplot()
    + theme_bw(base_size=18)
    + geom_line(d, aes('X', 'Y'), size=0.3, alpha=0.4)
    + geom_ribbon(d_est, aes('X', ymin='2.5%', ymax='97.5%'), fill='black', alpha=1/6)
    + geom_ribbon(d_est, aes('X', ymin='25%', ymax='75%'), fill='black', alpha=2/6)
    + geom_line(d_est, aes('X', y='50%'), size=0.2)
    + geom_line(d_est, aes('X', y='2.5%'), size=0.2)
    + geom_line(d_est, aes('X', y='97.5%'), size=0.2)
    + geom_line(d_est, aes('X', y='25%'), size=0.2)
    + geom_line(d_est, aes('X', y='75%'), size=0.2)
    + labs(x='Time (Second)', y='Y')
)
p.save(filename='output/fig11-9-left.py.png', dpi=300, width=3.5, height=3)


fit = cmdstanpy.from_csv('output/result-model11-7')
sw_ms = fit.stan_variable(var='sw')
qua = np.quantile(sw_ms, [0.025, 0.25, 0.50, 0.75, 0.975], axis=0)
d_est = pandas.DataFrame(np.column_stack([np.arange(1, T+1), qua.T]), \
    columns=['X', '2.5%', '25%', '50%', '75%', '97.5%'])

p = (ggplot()
    + theme_bw(base_size=18)
    + geom_line(d, aes('X', 'Y'), size=0.3, alpha=0.4)
    + geom_ribbon(d_est, aes('X', ymin='2.5%', ymax='97.5%'), fill='black', alpha=1/6)
    + geom_ribbon(d_est, aes('X', ymin='25%', ymax='75%'), fill='black', alpha=2/6)
    + geom_line(d_est, aes('X', y='50%'), size=0.2)
    + geom_line(d_est, aes('X', y='2.5%'), size=0.2)
    + geom_line(d_est, aes('X', y='97.5%'), size=0.2)
    + geom_line(d_est, aes('X', y='25%'), size=0.2)
    + geom_line(d_est, aes('X', y='75%'), size=0.2)
    + labs(x='Time (Second)', y='Y')
)
p.save(filename='output/fig11-9-right.py.png', dpi=300, width=3.5, height=3)

File: chap11/fig11-9.R
library(dplyr)
library(ggplot2)

d <- read.csv('input/data-switch.csv')
T <- nrow(d)

fit <- readRDS('output/result-model11-6.RDS')
sw_ms <- fit$draws('sw', format='matrix')
qua <- apply(sw_ms, 2, quantile, probs=c(0.025, 0.25, 0.50, 0.75, 0.975))
d_est <- data.frame(X=1:T, t(qua), check.names=FALSE)

p <- ggplot() +
  theme_bw(base_size=18) +
  geom_line(data=d, aes(x=X, y=Y), linewidth=0.3, alpha=0.4) +
  geom_ribbon(data=d_est, aes(x=X, ymin=`2.5%`, ymax=`97.5%`), fill='black', alpha=1/6) +
  geom_ribbon(data=d_est, aes(x=X, ymin=`25%`, ymax=`75%`), fill='black', alpha=2/6) +
  geom_line(data=d_est, aes(x=X, y=`50%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`2.5%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`97.5%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`25%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`75%`), linewidth=0.2) +
  labs(x='Time (Second)', y='Y')
ggsave(p, file='output/fig11-9-left.png', dpi=300, w=3.5, h=3)


fit <- readRDS('output/result-model11-7.RDS')
sw_ms <- fit$draws('sw', format='matrix')
qua <- apply(sw_ms, 2, quantile, probs=c(0.025, 0.25, 0.50, 0.75, 0.975))
d_est <- data.frame(X=1:T, t(qua), check.names=FALSE)

p <- ggplot() +
  theme_bw(base_size=18) +
  geom_line(data=d, aes(x=X, y=Y), linewidth=0.3, alpha=0.4) +
  geom_ribbon(data=d_est, aes(x=X, ymin=`2.5%`, ymax=`97.5%`), fill='black', alpha=1/6) +
  geom_ribbon(data=d_est, aes(x=X, ymin=`25%`, ymax=`75%`), fill='black', alpha=2/6) +
  geom_line(data=d_est, aes(x=X, y=`50%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`2.5%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`97.5%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`25%`), linewidth=0.2) +
  geom_line(data=d_est, aes(x=X, y=`75%`), linewidth=0.2) +
  labs(x='Time (Second)', y='Y')
ggsave(p, file='output/fig11-9-right.png', dpi=300, w=3.5, h=3)
